File: sample-chat-java/app/src/main/java/com/quickblox/sample/chat/java/ui/activity/MessageInfoActivity.java
package com.quickblox.sample.chat.java.ui.activity;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.widget.ListView;

import com.quickblox.chat.QBChatService;
import com.quickblox.chat.listeners.QBMessageStatusListener;
import com.quickblox.chat.model.QBChatMessage;
import com.quickblox.core.QBEntityCallback;
import com.quickblox.core.exception.QBResponseException;
import com.quickblox.sample.chat.java.R;
import com.quickblox.sample.chat.java.ui.adapter.UsersAdapter;
import com.quickblox.sample.chat.java.utils.ToastUtils;
import com.quickblox.sample.chat.java.utils.chat.ChatHelper;
import com.quickblox.sample.chat.java.utils.qb.QbUsersHolder;
import com.quickblox.users.QBUsers;
import com.quickblox.users.model.QBUser;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.Nullable;

public class MessageInfoActivity extends BaseActivity implements QBMessageStatusListener {
    private static final String TAG = MessageInfoActivity.class.getSimpleName();

    private static final String EXTRA_MESSAGE = "extra_message";
    private static final String EXTRA_MSG_INFO_TYPE = "extra_message_info_type";
    public static final String MESSAGE_INFO_DELIVERED_TO = "delivered_to";
    public static final String MESSAGE_INFO_READ_BY = "read_by";

    private QBChatMessage chatMessage;
    private String messageInfoType;
    private ListView usersListView;
    private ArrayList<QBUser> deliveredUsers = new ArrayList<>();
    private ArrayList<QBUser> readUsers = new ArrayList<>();

    public static void start(Context context, QBChatMessage chatMessage, String messageInfoType) {
        Intent intent = new Intent(context, MessageInfoActivity.class);
        intent.putExtra(EXTRA_MESSAGE, chatMessage);
        intent.putExtra(EXTRA_MSG_INFO_TYPE, messageInfoType);
        context.startActivity(intent);
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_chat_info);

        chatMessage = (QBChatMessage) getIntent().getSerializableExtra(EXTRA_MESSAGE);
        messageInfoType = getIntent().getStringExtra(EXTRA_MSG_INFO_TYPE);
        usersListView = findViewById(R.id.list_chat_info_users);
        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }

        if (messageInfoType.equals(MESSAGE_INFO_DELIVERED_TO)) {
            fillByDeliveredUsers();
        } else if (messageInfoType.equals(MESSAGE_INFO_READ_BY)) {
            fillByReadUsers();
        }
    }

    @Override
    public void onResumeFinished() {
        super.onResumeFinished();
        try {
            QBChatService.getInstance().getMessageStatusesManager().addMessageStatusListener(this);
        } catch (Exception e) {
            Log.d(TAG, e.getMessage());
            finish();
        }
    }

    @Override
    protected void onPause() {
        super.onPause();
        QBChatService.getInstance().getMessageStatusesManager().removeMessageStatusListener(this);
    }

    private void fillByDeliveredUsers() {
        if (getSupportActionBar() != null) {
            getSupportActionBar().setTitle(R.string.message_info_delivered_to);
        }
        loadDeliveredUsers();
    }

    private void fillByReadUsers() {
        if (getSupportActionBar() != null) {
            getSupportActionBar().setTitle(R.string.message_info_read_by);
        }
        loadReadUsers();
    }

    private void loadDeliveredUsers() {
        final List<Integer> deliveredIDs = new ArrayList<>(chatMessage.getDeliveredIds());

        ChatHelper.getInstance().getUsersFromMessage(chatMessage, new QBEntityCallback<ArrayList<QBUser>>() {
            @Override
            public void onSuccess(ArrayList<QBUser> qbUsers, Bundle bundle) {
                if (qbUsers != null && qbUsers.size() > 0) {
                    for (QBUser user : qbUsers) {
                        if (deliveredIDs.contains(user.getId())) {
                            deliveredUsers.add(user);
                        }
                    }
                }
                if (getSupportActionBar() != null) {
                    getSupportActionBar().setSubtitle(makeSubtitle(deliveredUsers.size()));
                }
                fillAdapter(deliveredUsers);
            }

            @Override
            public void onError(QBResponseException e) {
                showErrorSnackbar(R.string.select_users_get_users_error, e, null);
            }
        });
    }

    private void loadReadUsers() {
        final List<Integer> readIDs = new ArrayList<>(chatMessage.getReadIds());

        ChatHelper.getInstance().getUsersFromMessage(chatMessage, new QBEntityCallback<ArrayList<QBUser>>() {
            @Override
            public void onSuccess(ArrayList<QBUser> qbUsers, Bundle bundle) {
                if (qbUsers != null && qbUsers.size() > 0) {
                    for (QBUser user : qbUsers) {
                        if (readIDs.contains(user.getId())) {
                            readUsers.add(user);
                        }
                    }
                }
                if (getSupportActionBar() != null) {
                    getSupportActionBar().setSubtitle(makeSubtitle(readUsers.size()));
                }
                fillAdapter(readUsers);
            }

            @Override
            public void onError(QBResponseException e) {
                showErrorSnackbar(R.string.select_users_get_users_error, e, null);
            }
        });
    }

    private String makeSubtitle(int usersSize) {
        String result = "";
        if (usersSize == 0) {
            result = getString(R.string.message_info_noone);
        } else if (usersSize == 1) {
            result = getString(R.string.message_info_single_user);
        } else {
            result = usersSize + " " + getString(R.string.message_info_multiple_users);
        }
        return result;
    }

    private void fillAdapter(ArrayList<QBUser> qbUsers) {
        usersListView.setAdapter(new UsersAdapter(this, qbUsers));
        if (messageInfoType.equals(MESSAGE_INFO_DELIVERED_TO) && getSupportActionBar() != null) {
            getSupportActionBar().setSubtitle(makeSubtitle(deliveredUsers.size()));
        }
        if (messageInfoType.equals(MESSAGE_INFO_READ_BY) && getSupportActionBar() != null) {
            getSupportActionBar().setSubtitle(makeSubtitle(readUsers.size()));
        }
    }

    @Override
    public void processMessageDelivered(String messageID, String dialogID, Integer userID) {
        if (messageInfoType.equals(MESSAGE_INFO_DELIVERED_TO) && dialogID.equals(chatMessage.getDialogId())
                && messageID.equals(chatMessage.getId()) && userID != null) {
            QBUser user = QbUsersHolder.getInstance().getUserById(userID);
            if (user != null) {
                deliveredUsers.add(user);
                fillAdapter(deliveredUsers);
            } else {
                QBUsers.getUser(userID).performAsync(new QBEntityCallback<QBUser>() {
                    @Override
                    public void onSuccess(QBUser qbUser, Bundle bundle) {
                        if (qbUser != null) {
                            deliveredUsers.add(qbUser);
                        }
                        fillAdapter(deliveredUsers);
                    }

                    @Override
                    public void onError(QBResponseException e) {
                        ToastUtils.shortToast(e.getMessage());
                        Log.d(TAG, e.getMessage());
                    }
                });
            }
        }
    }

    @Override
    public void processMessageRead(String messageID, String dialogID, Integer userID) {
        if (messageInfoType.equals(MESSAGE_INFO_READ_BY) && dialogID.equals(chatMessage.getDialogId())
                && messageID.equals(chatMessage.getId()) && userID != null) {
            QBUser user = QbUsersHolder.getInstance().getUserById(userID);
            if (user != null) {
                readUsers.add(user);
                fillAdapter(readUsers);
            } else {
                QBUsers.getUser(userID).performAsync(new QBEntityCallback<QBUser>() {
                    @Override
                    public void onSuccess(QBUser qbUser, Bundle bundle) {
                        if (qbUser != null) {
                            readUsers.add(qbUser);
                        }
                        fillAdapter(readUsers);
                    }

                    @Override
                    public void onError(QBResponseException e) {
                        ToastUtils.shortToast(e.getMessage());
                        Log.d(TAG, e.getMessage());
                    }
                });
            }
        }
    }
}

File: sample-chat-kotlin/app/src/main/java/com/quickblox/sample/chat/kotlin/ui/activity/MessageInfoActivity.kt
package com.quickblox.sample.chat.kotlin.ui.activity

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.widget.ListView
import com.quickblox.chat.QBChatService
import com.quickblox.chat.listeners.QBMessageStatusListener
import com.quickblox.chat.model.QBChatMessage
import com.quickblox.core.QBEntityCallback
import com.quickblox.core.exception.QBResponseException
import com.quickblox.sample.chat.kotlin.R
import com.quickblox.sample.chat.kotlin.ui.adapter.UsersAdapter
import com.quickblox.sample.chat.kotlin.utils.chat.ChatHelper
import com.quickblox.sample.chat.kotlin.utils.qb.QbUsersHolder
import com.quickblox.sample.chat.kotlin.utils.shortToast
import com.quickblox.users.QBUsers
import com.quickblox.users.model.QBUser
import java.util.*

private const val EXTRA_MESSAGE = "extra_message"
private const val EXTRA_MSG_INFO_TYPE = "extra_message_info_type"
const val MESSAGE_INFO_DELIVERED_TO = "delivered_to"
const val MESSAGE_INFO_READ_BY = "read_by"

class MessageInfoActivity : BaseActivity(), QBMessageStatusListener {
    private val TAG = MessageInfoActivity::class.java.simpleName

    private lateinit var chatMessage: QBChatMessage
    private lateinit var messageInfoType: String
    private lateinit var usersListView: ListView
    private val deliveredUsers = ArrayList<QBUser>()
    private val readUsers = ArrayList<QBUser>()

    companion object {
        fun start(context: Context, chatMessage: QBChatMessage, messageInfoType: String) {
            val intent = Intent(context, MessageInfoActivity::class.java)
            intent.putExtra(EXTRA_MESSAGE, chatMessage)
            intent.putExtra(EXTRA_MSG_INFO_TYPE, messageInfoType)
            context.startActivity(intent)
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_chat_info)

        chatMessage = intent.getSerializableExtra(EXTRA_MESSAGE) as QBChatMessage
        messageInfoType = intent.getStringExtra(EXTRA_MSG_INFO_TYPE)
        usersListView = findViewById(R.id.list_chat_info_users)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        if (messageInfoType == MESSAGE_INFO_DELIVERED_TO) {
            fillByDeliveredUsers()
        } else if (messageInfoType == MESSAGE_INFO_READ_BY) {
            fillByReadUsers()
        }
    }

    override fun onResumeFinished() {
        super.onResumeFinished()
        try {
            QBChatService.getInstance().messageStatusesManager.addMessageStatusListener(this)
        } catch (e: Exception) {
            e.message?.let { Log.d(TAG, it) }
            finish()
        }
    }

    override fun onPause() {
        super.onPause()
        if (QBChatService.getInstance().messageStatusesManager != null) {
            QBChatService.getInstance().messageStatusesManager.removeMessageStatusListener(this)
        }
    }

    private fun fillByDeliveredUsers() {
        supportActionBar?.title = getString(R.string.message_info_delivered_to)
        loadDeliveredUsers()
    }

    private fun fillByReadUsers() {
        supportActionBar?.title = getString(R.string.message_info_read_by)
        loadReadUsers()
    }

    private fun loadDeliveredUsers() {
        val deliveredIDs = chatMessage.deliveredIds

        ChatHelper.getUsersFromMessage(chatMessage, object : QBEntityCallback<ArrayList<QBUser>> {
            override fun onSuccess(qbUsers: ArrayList<QBUser>?, b: Bundle?) {
                qbUsers?.let {
                    for (user in it) {
                        if (deliveredIDs.contains(user.id)) {
                            deliveredUsers.add(user)
                        }
                    }
                }
                supportActionBar?.subtitle = makeSubtitle(deliveredUsers.size)
                fillAdapter(deliveredUsers)
            }

            override fun onError(e: QBResponseException?) {
                showErrorSnackbar(R.string.select_users_get_users_error, e, null)
            }
        })
    }

    private fun loadReadUsers() {
        val readIDs = chatMessage.readIds

        ChatHelper.getUsersFromMessage(chatMessage, object : QBEntityCallback<ArrayList<QBUser>> {
            override fun onSuccess(qbUsers: ArrayList<QBUser>?, b: Bundle?) {
                qbUsers?.let {
                    for (user in it) {
                        if (readIDs.contains(user.id)) {
                            readUsers.add(user)
                        }
                    }
                }
                supportActionBar?.subtitle = makeSubtitle(readUsers.size)
                fillAdapter(readUsers)
            }

            override fun onError(e: QBResponseException?) {
                showErrorSnackbar(R.string.select_users_get_users_error, e, null)
            }
        })
    }

    private fun makeSubtitle(usersSize: Int): String {
        var result = ""
        if (usersSize == 0) {
            result = getString(R.string.message_info_noone)
        } else if (usersSize == 1) {
            result = getString(R.string.message_info_single_user)
        } else {
            result = usersSize.toString() + " " + getString(R.string.message_info_multiple_users)
        }
        return result
    }

    private fun fillAdapter(qbUsers: ArrayList<QBUser>) {
        usersListView.adapter = UsersAdapter(this, qbUsers as MutableList<QBUser>)
        if (messageInfoType == MESSAGE_INFO_DELIVERED_TO && supportActionBar != null) {
            supportActionBar!!.subtitle = makeSubtitle(deliveredUsers.size)
        }
        if (messageInfoType == MESSAGE_INFO_READ_BY && supportActionBar != null) {
            supportActionBar!!.subtitle = makeSubtitle(readUsers.size)
        }
    }

    override fun processMessageDelivered(messageID: String?, dialogID: String?, userID: Int?) {
        if (messageInfoType == MESSAGE_INFO_DELIVERED_TO && dialogID == chatMessage.dialogId
                && messageID == chatMessage.id && userID != null) {
            val user = QbUsersHolder.getUserById(userID)
            if (user != null) {
                deliveredUsers.add(user)
                fillAdapter(deliveredUsers)
            } else {
                QBUsers.getUser(userID).performAsync(object : QBEntityCallback<QBUser> {
                    override fun onSuccess(qbUser: QBUser?, b: Bundle?) {
                        qbUser?.let {
                            deliveredUsers.add(qbUser)
                        }
                        fillAdapter(deliveredUsers)
                    }

                    override fun onError(e: QBResponseException?) {
                        shortToast(e?.message)
                        Log.d(TAG, e?.message)
                    }
                })
            }
        }
    }

    override fun processMessageRead(messageID: String?, dialogID: String?, userID: Int?) {
        if (messageInfoType == MESSAGE_INFO_READ_BY && dialogID == chatMessage.dialogId
                && messageID == chatMessage.id && userID != null) {
            val user = QbUsersHolder.getUserById(userID)
            if (user != null) {
                readUsers.add(user)
                fillAdapter(readUsers)
            } else {
                QBUsers.getUser(userID).performAsync(object : QBEntityCallback<QBUser> {
                    override fun onSuccess(qbUser: QBUser?, b: Bundle?) {
                        qbUser?.let {
                            readUsers.add(qbUser)
                        }
                        fillAdapter(readUsers)
                    }

                    override fun onError(e: QBResponseException?) {
                        shortToast(e?.message)
                        Log.d(TAG, e?.message)
                    }
                })
            }
        }
    }
}